include/nbl/asset/filters/CFlattenRegionsStreamHashImageFilter.h

```diff
@@ -0,0 +1,220 @@
+					return false;
+			}
+
+			return true;
+		}
+
+		template<class ExecutionPolicy>
+		static inline bool execute(ExecutionPolicy&& policy, state_type* state)
+		{
+			if (!validate(state))
+				return false;
+
+			struct
+			{
+				CFlattenRegionsImageFilter::state_type flatten;			// CFlattenRegionsImageFilter's state, we use scratch memory to respecify regions & prefill with 0-value non-covered texels
+			} proxy;
+
+			/*
+				first we need to ensure that texels which are not covered by regions 
+				are filled with 0 value, flatten will also respecify regions to make
+				sure we don't have overlaping regions - we will override them anyway
+			*/
+
+			{
+				auto& flatten = proxy.flatten;
+				flatten.inImage = state->inImage;
+				flatten.outImage = state->scratchMemory.flatten;
+				flatten.preFill = true;
+				memset(flatten.fillValue.pointer, 0, sizeof(flatten.fillValue.pointer));
+
+				assert(CFlattenRegionsImageFilter::execute(policy, &proxy.flatten)); // this should never fail, at this point we are already validated
+			}
+
+			/*
+				now when the output is prepared we ignore respecified regions and go
+				with single region covering all texels for a given mip level & layers 
+			*/
+
+			const auto& parameters = proxy.flatten.outImage->getCreationParameters();
+			const uint8_t* inData = reinterpret_cast<const uint8_t*>(proxy.flatten.outImage->getBuffer()->getPointer());
+			const TexelBlockInfo info(parameters.format);
+			const auto bytesPerPixel = proxy.flatten.outImage->getBytesPerPixel();
+			const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(parameters.format);
+
+			// override regions, we need to cover all texels
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(parameters.mipLevels);
+			{
+				size_t bufferSize = 0ull;
+
+				for (auto rit = regions->begin(); rit != regions->end(); rit++)
+				{
+					auto miplevel = static_cast<uint32_t>(std::distance(regions->begin(), rit));
+					auto localExtent = proxy.flatten.outImage->getMipSize(miplevel);
+					rit->bufferOffset = bufferSize;
+					rit->bufferRowLength = localExtent.x; // could round up to multiple of 8 bytes in the future
+					rit->bufferImageHeight = localExtent.y;
+					rit->imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u);
+					rit->imageSubresource.mipLevel = miplevel;
+					rit->imageSubresource.baseArrayLayer = 0u;
+					rit->imageSubresource.layerCount = parameters.arrayLayers;
+					rit->imageOffset = { 0u,0u,0u };
+					rit->imageExtent = { localExtent.x,localExtent.y,localExtent.z };
+					auto levelSize = info.roundToBlockSize(localExtent);
+					auto memsize = levelSize[0] * levelSize[1] * levelSize[2] * parameters.arrayLayers * bytesPerPixel;
+
+					assert(memsize.getNumerator() % memsize.getDenominator() == 0u);
+					bufferSize += memsize.getIntegerApprox();
+				}
+			}
```

Comment:
this needs to happen in `allocateScratchMemory`

Reply:
done